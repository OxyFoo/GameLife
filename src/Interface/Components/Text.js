import * as React from 'react';
import { StyleSheet, Text as RNText, TouchableOpacity } from 'react-native';

import themeManager from '../../Managers/ThemeManager';

import { IsUndefined } from '../../Functions/Functions';

const MAIN_FONT_NAME = 'Hind Vadodara';

const TextProps = {
    style: {},
    bold: undefined,
    containerStyle: {},
    fontSize: 18,
    color: 'primary',
    onPress: undefined,
    onLayout: undefined
}

class Text extends React.Component {
    render() {
        const onPress = this.props.onPress;
        let color = themeManager.GetColor(this.props.color, 'text');
        if (color === null) {
            color = themeManager.GetColor(this.props.color);
        }

        return (
            <TouchableOpacity
                style={this.props.containerStyle}
                onPress={onPress}
                activeOpacity={.5}
                disabled={IsUndefined(onPress)}
            >
                <RNText
                    style={[
                        styles.text,
                        {
                            color: color,
                            fontSize: this.props.fontSize,
                            fontWeight: IsUndefined(this.props.bold) ? 'normal' : 'bold'
                        },
                        this.props.style
                    ]}
                    onLayout={this.props.onLayout}
                >
                    {this.props.children}
                </RNText>
            </TouchableOpacity>
        );
    }
}

Text.prototype.props = TextProps;
Text.defaultProps = TextProps;

const styles = StyleSheet.create({
    text: {
        margin: 0,
        padding: 0,
        textAlign: 'center',
        fontFamily: MAIN_FONT_NAME
    }
});

export default Text;
export { MAIN_FONT_NAME };