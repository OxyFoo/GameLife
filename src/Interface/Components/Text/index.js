import * as React from 'react';
import { StyleSheet, Text as RNText, TouchableOpacity } from 'react-native';

import themeManager from 'Managers/ThemeManager';

const MAIN_FONT_NAME = 'Hind Vadodara';

/**
 * @typedef {import('react-native').TextStyle} TextStyle
 * @typedef {import('react-native').ViewStyle} ViewStyle
 * @typedef {import('react-native').StyleProp<ViewStyle | TextStyle>} TextStyleProp
 * @typedef {import('react-native').StyleProp<ViewStyle>} ViewStyleProp
 * @typedef {import('react-native').TextProps} TextProps
 * @typedef {import('react-native').StyleProp<TextStyle>} StyleProp
 * 
 * @typedef {import('Managers/ThemeManager').ThemeText} ThemeText
 * @typedef {import('Managers/ThemeManager').ThemeColor} ThemeColor
 * 
 * @typedef {Object} TextPropsType
 * @property {TextStyleProp} style
 * @property {ViewStyleProp} containerStyle Style of touchable opacity for onPress text
 * @property {number} fontSize
 * @property {ThemeColor | ThemeText} color
 */

/** @type {TextProps & TextPropsType} */
const TextProps = {
    style: {},
    containerStyle: {},
    fontSize: 18,
    color: 'primary'
};

class Text extends React.Component {
    /** @param {TextProps & TextPropsType} nextProps */
    shouldComponentUpdate(nextProps) {
        return this.props.children !== nextProps.children ||
            this.props.color !== nextProps.color ||
            this.props.fontSize !== nextProps.fontSize;
    }

    render() {
        const {
            style, containerStyle, color, fontSize, onPress, children, ...props
        } = this.props;

        if (typeof children !== 'string') {
            return null;
        }

        /** @type {StyleProp} */
        const textStyle = {
            ...styles.text,
            color: themeManager.GetColor(color),
            fontSize: fontSize
        };

        if (!!onPress) {
            return (
                <TouchableOpacity
                    style={containerStyle}
                    onPress={onPress}
                    activeOpacity={.5}
                >
                    <RNText style={[textStyle, style]} {...props}>
                        {children}
                    </RNText>
                </TouchableOpacity>
            );
        }

        return (
            <RNText style={[textStyle, style]} {...props}>
                {children}
            </RNText>
        );
    }
}

Text.prototype.props = TextProps;
Text.defaultProps = TextProps;

const styles = StyleSheet.create({
    text: {
        margin: 0,
        padding: 0,
        textAlign: 'center',
        fontFamily: MAIN_FONT_NAME
    }
});

export { Text, MAIN_FONT_NAME };
