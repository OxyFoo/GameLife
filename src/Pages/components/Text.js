import * as React from 'react';
import { StyleSheet, Text as RNText, TouchableOpacity } from 'react-native';

import themeManager from '../../Managers/ThemeManager';
import { isUndefined } from '../../Functions/Functions';

const MAIN_FONT_NAME = 'Hind Vadodara';

const TextProps = {
    style: {},
    containerStyle: {},
    fontSize: 18,
    color: 'primary',
    onPress: undefined,
}

class Text extends React.Component {
    render() {
        const color = themeManager.getColor(this.props.color, 'text');
        const onPress = this.props.onPress;

        return (
            <TouchableOpacity
                style={this.props.containerStyle}
                onPress={onPress}
                activeOpacity={.5}
                disabled={isUndefined(onPress)}
            >
                <RNText
                    style={[
                        styles.text,
                        {
                            color: color,
                            fontSize: this.props.fontSize
                        },
                        this.props.style
                    ]}
                >
                    {this.props.children}
                </RNText>
            </TouchableOpacity>
        );
    }
}

Text.prototype.props = TextProps;
Text.defaultProps = TextProps;

const styles = StyleSheet.create({
    text: {
        margin: 0,
        padding: 0,
        textAlign: 'center',
        fontFamily: MAIN_FONT_NAME
    }
});

export default Text;
export { MAIN_FONT_NAME };