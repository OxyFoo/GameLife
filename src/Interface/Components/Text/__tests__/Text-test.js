import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

// Graphical components
import Text from '../index';

describe('[Component] Text', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<Text />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render the correct text content', () => {
        const textContent = 'Hello, World!';
        const { getByText } = render(<Text>{textContent}</Text>);
        const textElement = getByText(textContent);

        expect(textElement).toBeDefined();
    });

    it('should have the correct styles applied', () => {
        const { getByText } = render(
            <Text style={{ color: 'red', fontSize: 16 }}>
                Styled Text
            </Text>
        );
        const textElement = getByText('Styled Text');

        // Get only user-defined styles
        const textStyle = textElement.props.style.at(-1);
        expect(textStyle).toEqual({ color: 'red', fontSize: 16 });
    });
});