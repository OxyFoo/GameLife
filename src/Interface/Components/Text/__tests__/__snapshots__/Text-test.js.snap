// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Text should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <Text
    onLayout={[Function]}
    style={
      [
        {
          "fontFamily": "Hind Vadodara",
          "margin": 0,
          "padding": 0,
          "textAlign": "center",
        },
        {
          "color": "#ECECEC",
          "fontSize": 18,
          "fontWeight": "normal",
        },
        {},
      ]
    }
  />
</View>
`;
